### ISSUE TITLE
Drop the S2325 rule which generates too many false positives

### ISSUE THREAD

**Issue by dbolkensteyn** [SonarSource/sonarlint-vs#65]
S2325 reports false positives on 
* test methods
* MVC and WebApi action methods
* Serializable and DataContract attributed properties.
* And basically on any method and property that needs to be reached through reflection, because most/many/several reflection based frameworks don't handle static properties/methods.

We could ignore 
* methods/properties that have attributes on them, 
* properties of classes with attributes
* methods of classes that derive from `Controller`.

But there will always be cases that report false positives, or result in false negatives. The overall gain from this rule is minimal, so considering these, we decided to drop the rule.

**Comment by OsirisTerje** [comment]
The important issue here is to not report on test Methods.  Otherwise I think this should report, regardless of attributes otherwise.
It should know about the common known testframeworks, MSTest, NUnit and XUnit and the attributes used there.  It can always add more as they appear. 
The attributes it need to check for is then, IIRC:
MSTest: TestMethod
NUnit: Test, TestCase, TestCaseSource
XUnit: Fact
In addition the setup and teardown attributes (XUnit uses ctor's so no issue there)
MSTest: ClassInitialize, ClassCleanup TestInitialize, TestCleanup
NUnit: (V2): TestFixtureSetup, TestFixtureTeardown, SetUp, TearDown (V3) OneTimeSetUp, OneTimeTearDown

**Comment by dbolkensteyn** [comment]
I see what you mean @OsirisTerje - however:

1. I prefer to keep the logic of the rule quite small and simple - and the same goes for the rule description. The fewer cases we have to deal with, the better.
2. False negatives should be preferred over false positives: I can easily imagine other non-testing frameworks to rely on attributes on methods, and require them to be non-static (I've heard that WCF has that).

Do you think that the rule will miss so many cases if we go forward this the issue as it is?

**Comment by OsirisTerje** [comment]
Yes, I think it will.  Attributes are used for so many different things, that having them all excluded from analysis doesn't feel right.  
Just think of the Serializable attribute, which is pretty common.  If all classes marked by that is excluded, people will be confused.

**Comment by dbolkensteyn** [comment]
The idea is to not report on a property/method if it has an attribute itself: We won't be looking at whether or not the class containing it has an attribute.

Indeed the `Serializable` attribute is a good example: According to this question: http://stackoverflow.com/questions/17222900/serialize-object-along-with-static-member-variables-to-xml - static properties will be ignored.

It might be quite tricky to find that out while debugging, so I think this is another argument in favor of ignoring as soon as any attribute is present?

**Comment by OsirisTerje** [comment]
Ok, so Serialize was not that good as an example, as it mostly deals With Properties, like XmlIgnore.
But, create a WebAPI Project, or any MVC Application, and you see Properties being used for a lot of the methods, example
  // GET api/Account/UserInfo
        [HostAuthentication(DefaultAuthenticationTypes.ExternalBearer)]
        [Route("UserInfo")]
        public UserInfoViewModel GetUserInfo()
        {

or

 [Route("ManageInfo")]
        public async Task<ManageInfoViewModel> GetManageInfo(string returnUrl, bool generateState = false)
        {

Do you think it is a good idea that these methods are not being analyzed ?

Attributes is a way of declaring behaviour, in an aspect oriented way, so to say that these should not be analyzed for possible static-ness feels very wrong.

**Comment by OsirisTerje** [comment]
I have another issue with this kind of warning.  The "Cure" is said to add the static keyword, but what it really indicates is that the Method doesnt really belong to the class where it is placed.  It should advise, - or preferably the code fix should suggest  - to move the Method to one of the classes in the parameter list - if any, before it advises to set the Method as static.

**Comment by tvsonar** [comment]
@OsirisTerje I think WebAPI and MVC action methods are also not good examples here, because they can't be made static (see http://stackoverflow.com/questions/8899098/can-we-create-static-actionresult-method-in-controller). 

You are right that attributes are an elegant way of adding behavior in an AOP way, but we couldn't find a real-world, widely-used case where you would want to put attributes on static methods. 

Also note that we wouldn't exclude the method or property from all the analysis, but only from this "member can be made static" one ([S2325](http://vs.sonarlint.org/rules/index.html#version=1.0.0&ruleId=S2325)).

**Comment by dbolkensteyn** [comment]
just something which might not be that obvious @OsirisTerje : We prefer false negatives over false positives.

A false negative is when the tool fails to report an actual issue. A false positive is when the tool reports an issue which is actually not real.

If a tool generates a massive amount of false positives, developers will start to distrust it, and ultimately uninstall the tool.

If we suggest to make a method or property `static` - but this later on breaks serialization and is only found by integration tests or worse by customers - it will have a huge negative impact.

Compared to that - the benefits of making the method `static` is actually really minor.

So to sum up: We want to exclude all properties or methods that have an attribute, or that are contained in a class which itself has an attribute.

**Comment by tvsonar** [comment]
The [first comment](https://github.com/SonarSource/sonarlint-vs/issues/65#issue-96320439) has been updated to summarize the reasons for dropping this rule.

**Comment by OsirisTerje** [comment]
Isn't it a bit far to drop the rule all together ?

**Comment by tvsonar** [comment]
@OsirisTerje we tried to find a good solution, but without having to maintain a list of libraries or exceptions we couldn't. This shows that the rule isn't generally applicable and that it'll generate false positives and loads of noise, which we try to avoid.
